File: articles/development-differentiation-making-of-organisms/python/developmental_growth.py
"""
Developmental growth: exponential fit, doubling time, and stage-dependent logistic growth.

Run:
    python python/developmental_growth.py
"""

from __future__ import annotations

from pathlib import Path

import numpy as np
import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
GROWTH_PATH = ARTICLE_DIR / "data" / "developmental_growth.csv"


def stage_dependent_growth(
    t_end: float = 40,
    dt: float = 0.1,
    N0: float = 1e4,
    K: float = 8e4,
    r_early: float = 0.08,
    r_late: float = 0.03,
    switch_time: float = 18,
) -> pd.DataFrame:
    """Simulate logistic growth with a developmental stage shift."""

    times = np.arange(0, t_end + dt, dt)
    cells = np.zeros_like(times)
    cells[0] = N0

    for i in range(1, len(times)):
        r = r_early if times[i - 1] < switch_time else r_late
        dN = r * cells[i - 1] * (1 - cells[i - 1] / K)
        cells[i] = cells[i - 1] + dN * dt

    return pd.DataFrame({"time": times, "cells": cells})


def main() -> None:
    """Estimate early growth and run stage-dependent simulation."""

    growth = pd.read_csv(GROWTH_PATH)
    early = growth.iloc[:5].copy()

    slope, intercept = np.polyfit(early["time_h"], np.log(early["cells"]), 1)

    r_est = slope
    N0_est = np.exp(intercept)
    doubling_time_h = np.log(2) / r_est

    summary = pd.DataFrame(
        {
            "r_est": [r_est],
            "N0_est": [N0_est],
            "doubling_time_h": [doubling_time_h],
        }
    )

    sim = stage_dependent_growth()

    print(summary.round(4).to_string(index=False))
    print(sim.head(12).round(2).to_string(index=False))
    print(sim.tail(12).round(2).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/development-differentiation-making-of-organisms/r/developmental_growth.R
# Developmental growth workflow in R.
#
# Fits early exponential growth and simulates constrained logistic growth.

library(dplyr)

growth_path <- file.path("data", "developmental_growth.csv")

if (!file.exists(growth_path)) {
  growth_path <- file.path("..", "data", "developmental_growth.csv")
}

growth_df <- read.csv(growth_path)

fit_exp <- lm(log(cells) ~ time_h, data = growth_df %>% slice(1:5))

r_est <- coef(fit_exp)[["time_h"]]
N0_est <- exp(coef(fit_exp)[["(Intercept)"]])
doubling_time_h <- log(2) / r_est

growth_summary <- tibble(
  r_est = r_est,
  N0_est = N0_est,
  doubling_time_h = doubling_time_h
)

print(growth_summary)

growth_df <- growth_df %>%
  mutate(exp_pred = exp(predict(fit_exp, newdata = growth_df)))

print(growth_df)

simulate_logistic <- function(times, N0 = 1e4, r = 0.07, K = 6.2e4) {
  N <- numeric(length(times))
  N[1] <- N0

  for (i in 2:length(times)) {
    dt <- times[i] - times[i - 1]
    dN <- r * N[i - 1] * (1 - N[i - 1] / K)
    N[i] <- N[i - 1] + dN * dt
  }

  tibble(time = times, logistic_N = N)
}

sim_df <- simulate_logistic(seq(0, 40, by = 0.25))

print(sim_df %>% slice_head(n = 12))
print(sim_df %>% slice_tail(n = 12))
